Change coordinate handling logic

roji opened the pull request:
This PR changes ordinate handling based on the discussion in #8, and as detailed below. @FObermaier I'm not sure if this is exactly what you intended, but please take a look (especially at the new test) to see the proposed behavior; if you guys see an issue with it I can revisit and amend. @YohDeadfall and @austindrenski your input would be great too.

* HandleOrdinates on both reader and writer now act only as an "upper bound" - coordinates not specified in HandleOrdinates will be stripped when reading and writing objects.
* The meaning of Ordinates.None has been changed to mean that objects are read and written as-is, without either stripping or adding any coordinates.
* Ordinates.None is now the default.

Fixes #8

airbreather commented on NetTopologySuite.IO.PostGis/PostGisReader.cs:
```diff
@@ -211,30 +211,36 @@ protected ICoordinateSequence ReadCoordinateSequenceRing(BinaryReader reader, IC
-            var sequence = factory.Create(numPoints, HandleOrdinates);
+            var outputOrdinates = receivedOrdinates;
+            if (HandleOrdinates != Ordinates.None)
+                outputOrdinates &= HandleOrdinates;
```
`HandleOrdinates` still allows any ol' subset of XYZM, which means that this would let us call `factory.Create(numPoints, Ordinates.Y | Ordinates.M)` (i.e., no `Ordinates.X` flag).

This isn't a new issue, but if we're going to be making some **very** related changes, I think it makes sense to fix it as part of this.

airbreather replied:
`PostGisWriter` has set a precedent that the setter should bitwise-OR the `Ordinates.XY` flag onto a non-`None` value as it comes in, and I see no fundamental reason why we shouldn't follow suit here (if anything, it seems to make **more** sense to do it in the reader than to do it in the writer).

roji replied:
Sure, makes sense to me, I'll have HandleOrdinates do an OR with Ordinates.XY just like PostGisWriter does (unless it's None of course).

roji replied:
Will add a test too.

airbreather commented on NetTopologySuite.IO.PostGis/PostGisWriter.cs:
```diff
@@ -170,20 +149,20 @@ private static void Write(ICoordinateSequence sequence, Ordinates ordinates, Bin
                 {
                     double m = sequence.GetOrdinate(i, Ordinate.M);
-                    if (double.IsNaN(m)) m = 0d;
+                    Debug.Assert(!double.IsNaN(m));
```
We should just skip these two `Debug.Assert` lines (this one, for M, and the other one, for Z).  Nobody's ever going to see it since we don't release debug builds, and it can trigger even if the rest of our code is correct (since it depends fully on the input data).

This is assuming that I'm not wrong in my understanding that `Debug.Assert` calls are meant to inform the developer that, if the rest of our code is correct, its condition will always be true at the point in time when the `Debug.Assert` call is made.

airbreather replied:
If I'm wrong and these `Debug.Assert` lines do in fact fulfill a purpose, then please use an overload such as [`Debug.Assert(bool, string)`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debug.assert?view=netframework-4.7.2#System_Diagnostics_Debug_Assert_System_Boolean_System_String_) that allows you to provide a message explaining in more detail what happened and, perhaps, why it's bad.

roji replied:
Yeah, this is mainly meant for failing tests, to alert us earlier and directly as to the broken assumption. I added an informative message and I think it has some value, but if you really feel that it's unnecessary let me know and I'll remove it.

airbreather replied:
> if you really feel that it's unnecessary let me know and I'll remove it.

Per [@peetw's testing](https://github.com/NetTopologySuite/NetTopologySuite.IO.PostGis/pull/9#discussion_r248848731), I think this (and its counterpart assert for Z) should just go away.  It sounds like we'd faithfully preserve whatever values are carried by the instance, in which case we don't really have much of a reason to fret about it at this layer of the system.

roji replied:
Agreed.

airbreather commented on NetTopologySuite.IO.PostGis/PostGisWriter.cs:
```diff
@@ -170,20 +149,20 @@ private static void Write(ICoordinateSequence sequence, Ordinates ordinates, Bin
                 {
                     double z = sequence.GetOrdinate(i, Ordinate.Z);
-                    if (double.IsNaN(z)) z = 0d;
+                    Debug.Assert(!double.IsNaN(z));
```
We should just skip these two `Debug.Assert` lines (this one, for Z, and the other one, for M).  Nobody's ever going to see it since we don't release debug builds, and it can trigger even if the rest of our code is correct (since it depends fully on the input data).

This is assuming that I'm not wrong in my understanding that `Debug.Assert` calls are meant to inform the developer that, if the rest of our code is correct, its condition will always be true at the point in time when the `Debug.Assert` call is made.

Especially for this one (Z), I expect to see some `NaN` values on a somewhat regular basis, in cases where someone just creates a new geometry with all the default services that are set up for you (at least as of NTS 1.15... I'm still not sure how 1.16 will look), since I'm pretty sure that a 3-dimensional coordinate sequence is still the default even though NTS is advertised as a library for handling primarily 2-dimensional geometry.

airbreather replied:
If I'm wrong and these `Debug.Assert` lines do in fact fulfill a purpose, then please use an overload such as [`Debug.Assert(bool, string)`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debug.assert?view=netframework-4.7.2#System_Diagnostics_Debug_Assert_System_Boolean_System_String_) that allows you to provide a message explaining in more detail what happened and, perhaps, why it's bad.

roji replied:
Do you mean to say that it's possible to encounter a point `p` which has `p.CoordinateSequence.Ordinates` equal to `XYZ`, but when I access its Z coordinate I get a NaN (unset or something)? If so then the current code definitely isn't right and needs to be fixed - but I'm really not sure if we can actually send anything to PostgreSQL that makes sense... I don't think it'd accept NaN (but can check), so probably raising an exception is the way to go.

Let me know...

airbreather replied:
> Do you mean to say that it's possible to encounter a point `p` which has `p.CoordinateSequence.Ordinates` equal to `XYZ`, but when I access its Z coordinate I get a NaN (unset or something)?

Yes, and it doesn't require anything esoteric to get into that situation either; someone who is unfamiliar with the idiosyncrasies of JTS / NTS will more likely than not stumble into it:
# ConsoleApp0.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.2</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="NetTopologySuite.Core" Version="1.15.1" />
  </ItemGroup>

</Project>
```

# Program.cs
```csharp
using System;
using GeoAPI.Geometries;
using NetTopologySuite.Geometries;

class Program
{
    static void Main()
    {
        var seq = new Point(0, 0).CoordinateSequence;
        Console.WriteLine(seq.Ordinates); // writes XYZ
        var z = seq.GetOrdinate(0, Ordinate.Z);
        Console.WriteLine(z); // writes NaN
    }
}
```

> I'm really not sure if we can actually send anything to PostgreSQL that makes sense... I don't think it'd accept NaN (but can check), so probably raising an exception is the way to go.

I don't have enough experience with the PostgreSQL / PostGIS side of things to be able to comment on that, but I suspect that this is a reason why the current version of the code fills in zero wherever it reads a NaN.

peetw replied:
> I don't have enough experience with the PostgreSQL / PostGIS side of things to be able to comment on that, but I suspect that this is a reason why the current version of the code fills in zero wherever it reads a NaN.

In PostGis, it is certainly possible to create a geometry that has NaN co-ordinate values. For example, creating a point with a NaN Z co-ord:

    SELECT ST_MakePoint(0, 0, 'NaN');

yields:

     010100008000000000000000000000000000000000000000000000F8FF

which suggests that PostGis uses `000000000000F8FF` (i.e. the standard IEEE 754 NaN) to represent NaN co-ordinate values. So it should be fine? Possibly no need even for the `Assert.Debug` statement?

roji replied:
Thanks for testing that @peetw, agreed! I've removed the assertions, NaNs will now be written to PostgreSQL just like any other value.

airbreather reviewed (changes requested):
I think this is not actually bad overall, despite the scary X that GitHub puts when I do a "Request Changes" vote (edit: apparently they've changed it since the last review I voted that way, which makes me happy)... see specific points I'd ask to be addressed in some way before merging.

I'd also like to see a response from @FObermaier before I move to merge it, since he had the stronger opinions on that thread (mine were just minor concerns about consistency, one of which even appears to have been unfounded after re-reviewing `SqlServerBytesReader`).

Thank you for your submission!

austindrenski commented on NetTopologySuite.IO.PostGis/PostGisReader.cs:
```diff
@@ -211,30 +211,36 @@ protected ICoordinateSequence ReadCoordinateSequenceRing(BinaryReader reader, IC
-                if (getM) ordinateM = reader.ReadDouble();
-                if (handleM) sequence.SetOrdinate(i, Ordinate.M, ordinateM);
+
+                if (receivedOrdinates.HasFlag(Ordinates.Z))
```
Does this library always target versions with the new intrinsics?

airbreather replied:
No, but I feel that any inefficiencies of `.HasFlag` relative to the bitwise operations should be insignificant compared to the I/O that happens throughout this loop.

airbreather replied:
That said, these conditionals are all loop-invariant, so it wouldn't hurt either readability **or** performance to keep the `getZ` / `getM` and `handleZ` / `handleM` variables where they used to be (maybe with slightly better names for the `handle` ones), and just change the logic to initialize them using `HasFlag`...

roji replied:
Perf-wise, I'm slowly switching to using `HasFlag()` with the assumption that modern .NET Core runtimes are being used. Note that this doesn't have anything to do with the library AFAIK, but rather with the user application using the library - if it's running a recent runtime you'll get the efficient JITting.

On the other hand you're right about extracting it out into a variable - I somehow missed that we were in a loop. Extracted them out again with hopefully better naming (`receivedZ`, `outputtingZ`).

YohDeadfall replied:
The optimization is implemented in Core CLR via dotnet/coreclr#13748 and doesn't depend on the targeted by the library framework.

airbreather replied:
> doesn't depend on the targeted by the library framework

Right; however, the reality is that this library supports NetFX, which, to my knowledge, still has a (comparatively) slow implementation of `System.Enum.HasFlag`.

Given that this method is just a more readable way to formulate a simple bitwise test that JITs to something that should be fast enough on **all** frameworks, there's a persuasive argument to be made that library developers in situations like ours should favor writing the bitwise test instead of calling that method, especially inside of loops (which this was).

I personally think it's rare that the difference will ever matter (in my experience, there's usually something very near the `.HasFlag` call that I'd expect to drown it out, and this time was no exception), but IMO it's a totally valid question to ask.

austindrenski commented on NetTopologySuite.IO.PostGis/PostGisWriter.cs:
```diff
@@ -43,43 +44,20 @@ public PostGisWriter(ByteOrder encodingType)
+            int coordinateSpace = 8 * OrdinatesUtility.OrdinatesToDimension(maxCoords);
             byte[] bytes = GetBytes(geometry, coordinateSpace);
-            Write(geometry, ordinates, new MemoryStream(bytes));
+            Write(geometry, new MemoryStream(bytes));
```
Wasn't there talk recently of a forthcoming poolable memory stream? Does that sound familiar to anyone else?

airbreather replied:
Are you referring to [`Microsoft.IO.RecyclableMemoryStream`](https://github.com/Microsoft/Microsoft.IO.RecyclableMemoryStream)?

Regardless, this actually strikes me as a use case where the built-in `MemoryStream` is perfectly fine, since it wraps a fixed-size buffer that **needs** to be a new instance in order to fulfill the contract of this method.

AFAIK, recycling / pooling `MemoryStream` replacements are primarily concerned with reducing the costs associated with allocating transient buffers (especially those big enough to go onto the LOH) as the stream fills up to an unknown size.

austindrenski replied:
Ah, that does looks like what I was remembering (albeit not quite so recent anymore). But your explanation makes sense, thanks!

roji replied:
I personally don't care much about the particular `Write()` overload which creates the MemoryStream, since when actual I/O is performed against PostgreSQL an actual network stream is used instead...

roji commented:
@airbreather pushed commits as per your review comments, but one important question is still open...

YohDeadfall commented on NetTopologySuite.IO.PostGis/PostGisReader.cs:
```diff
@@ -211,30 +211,41 @@ protected ICoordinateSequence ReadCoordinateSequenceRing(BinaryReader reader, IC
+                {
+                    var z = reader.ReadDouble();
+                    if (outputtingZ)
+                        sequence.SetOrdinate(i, Ordinate.Z, z);
```
The precision model must be used here and below.
```suggestion
                        sequence.SetOrdinate(i, Ordinate.Z, precisionModel.MakePrecise(z));
```

FObermaier replied:
No, `NetTopologySuite.IO.WKBReader` and `NetTopologySuite.IO.WKTReader` only apply precisison model to x- and y-ordinates:
https://github.com/NetTopologySuite/NetTopologySuite/blob/583ae18b593e5835630aa7e67bb82dd8eb349ccd/NetTopologySuite/IO/WKBReader.cs#L272-L311

This is conformant with JTS.

YohDeadfall replied:
Is compatibility so important? In such cases it sounds like a bug which should be fixed in the two projects.

airbreather replied:
> it sounds like a bug

I don't consider this to be a bug.  AFAIK, JTS / NTS only "really" use the X and Y ordinates for its own operations and Z/M are just sorta... "there".

It's possible that perhaps (often? [usually?]) Z/M are at a different scale than X/Y, and so the same `PrecisionModel` might not even be appropriate for them.

roji replied:
I tend to agree with @YohDeadfall that it's strange and quite error-prone to treat Z/M differently in any way from X/Y, regardless of whether JTS/NTS perform operations on them or not... Even if Z/M may have different scales, it doesn't seem like that's necessarily the case, and support is needed for different precision models for different ordinates, then that's a new feature...

However, you guys are the GIS experts and in any case this is out-of-scope from the changes of this PR... We can continue discussion in a new issue.

airbreather replied:
> We can continue discussion in a new issue

I've opened NetTopologySuite/NetTopologySuite#280 for this.

YohDeadfall commented on NetTopologySuite.IO.PostGis/PostGisReader.cs:
```diff
@@ -211,30 +211,41 @@ protected ICoordinateSequence ReadCoordinateSequenceRing(BinaryReader reader, IC
+                {
+                    var m = reader.ReadDouble();
+                    if (outputtingM)
+                        sequence.SetOrdinate(i, Ordinate.M, m);
```
```suggestion
                        sequence.SetOrdinate(i, Ordinate.M, precisionModel.MakePrecise(m));
```

FObermaier replied:
dito.

airbreather commented:
> @airbreather pushed commits as per your review comments, but one important question is still open...

[Responded](https://github.com/NetTopologySuite/NetTopologySuite.IO.PostGis/pull/9#discussion_r249039830) on one of them, but otherwise I'm fine with this once that one question is resolved.

Note that I did merge #7 (after you'd originally submitted this, sorry for the conflicts), so I think one of your interim commits (b27ac38) may be redundant.

roji committed:
Change coordinate handling logic

* HandleOrdinates on both reader and writer now act only as an "upper
  bound" - coordinates not specified in HandleOrdinates will be
  stripped when reading and writing objects.
* The meaning of Ordinates.None has been changed to mean that objects
  are read and written as-is, without either stripping or adding any
  coordinates.
* Ordinates.None is now the default.

Fixes #8

roji commented:
@airbreather I rebased on the latest master, incorporating #7 and resolving the conflict. I also squashed all commits - let me know if there's anything else you'd like to have changed or discussed.

airbreather commented:
> @airbreather I rebased on the latest master, incorporating #7 and resolving the conflict. I also squashed all commits - let me know if there's anything else you'd like to have changed or discussed.

From a brief look, I think it looks great -- I'll double-check tonight.

@FObermaier do you have any lingering concerns?  If not, I intend to merge it tonight unless my "double-check" reveals a compelling reason not to.

FObermaier commented:
@airbreather I have no objections to merging the PR.
@roij et. al., thanks for your input.

airbreather commented:
> unless my "double-check" reveals a compelling reason not to.

...well... turns out `net35-client` doesn't have `Enum.HasFlag`, so the solution won't actually build for that target framework.

Edit: really quick extension method fixed it right up, so I've added a commit on top and merged.  Thanks again!

austindrenski commented:
> ...well... turns out `net35-client` doesn't have `Enum.HasFlag`, so the solution won't actually build for that target framework.

~~What's the support status of `net35-client`?~~ 
Nevermind, Google says 2028.

airbreather commented:
> > ...well... turns out `net35-client` doesn't have `Enum.HasFlag`, so the solution won't actually build for that target framework.
> 
> What's the support status of `net35-client`?

My preference would be to make the NTS libraries `netstandard2.0`-only starting with the next major version, but otherwise keep support for other target frameworks in place as long as it's not adding **too much** friction to do so.
